### PR TITLE
Materialize ancestor information

### PR DESCRIPTION
**Pull request by ctron**
~~**Note:** this PR builds on top of https://github.com/guacsec/trustify/pull/2407, only the last few commits are relevant.~~

It creates a table materializing the ancestor information of the RH model. Thus saving us from re-exploring that over and over again.

This brings down the "pypi request" query from days to ~13 mins:

```
➜  trustify git:(feature/atlas_perf_1_mat) time xh GET localhost:8080/api/v2/analysis/latest/component?q=purl=pkg%3Apypi%2Frequests (oidc token -Hf trusty)
HTTP/1.1 200 OK
Access-Control-Allow-Credentials: true
Access-Control-Expose-Headers: content-type
Content-Encoding: br
Content-Type: application/json
Date: Fri, 19 Jun 2026 06:49:53 GMT
Transfer-Encoding: chunked
Vary: Origin, Access-Control-Request-Method, Access-Control-Request-Headers
Vary: accept-encoding

{
    "items": [],
    "total": 0
}



________________________________________________________
Executed in  747.02 secs      fish           external
   usr time   24.16 millis    0.00 micros   24.16 millis
   sys time   20.16 millis  929.00 micros   19.23 millis

```

However, this is a draft and it may break stuff. So let's be optimistic, but cautious.

## Summary by Sourcery

Materialize SBOM ancestor and describing-CPE relationships to avoid expensive runtime graph walks and improve analysis query performance.

New Features:
- Introduce materialized sbom_describing_cpe and sbom_ancestor tables plus corresponding ORM entities to store SBOM describing CPEs and cross-SBOM ancestor links.
- Populate describing CPEs and ancestor links during SBOM ingestion for CycloneDX and SPDX documents.
- Add a Containerfile and .containerignore for building and packaging the trustd binary into a container image.

Enhancements:
- Refactor SBOM CPE and ancestor resolution to use batched queries and the new materialized tables instead of recursive in-memory graph traversal.
- Optimize product advisory and context CPE queries to read from sbom_describing_cpe, simplifying SQL and reducing join cost.
- Improve logging with truncated iterable formatting and richer instrumentation for graph loading and ranking.
- Add a helper to migrate the database up to a specific migration and extend the trustd db CLI to support targeted migrations.
- Update several dependencies (moka, petgraph, sea-orm, tokio, uuid) and align uuid usage in the storage module.
- Adjust tests and entities to reflect the new ancestor and describing-CPE materialization behavior.

Tests:
- Add tests verifying sbom_describing_cpe semantics and that ingesting linked SBOMs correctly populates the sbom_ancestor table.

## Summary by Sourcery

Materialize cross-SBOM ancestor relationships and tighten read-only transaction handling to speed up analysis queries and avoid connection pool exhaustion.

New Features:
- Introduce a materialized sbom_ancestor table and ORM entity to store cross-SBOM ancestor links between SBOMs.
- Populate sbom_ancestor entries during CycloneDX and SPDX SBOM ingestion based on checksum-linked external nodes.

Enhancements:
- Refactor SBOM CPE resolution to read ancestor CPEs via sbom_ancestor and sbom_describing_cpe instead of on-the-fly graph walking and checksum-based DFS.
- Centralize the begin_read helper in the common db module and adopt consistent read-only transactions across analysis, SBOM, advisory, product, purl, vulnerability, organization, and license endpoints.
- Improve graph loading logging with truncated iterator formatting and minor utility cleanups (e.g., timestamp helper, error wiring).

Tests:
- Add tests ensuring sbom_ancestor entries are populated for linked SBOMs and that deep descendant queries only use a single lazy pool connection.
- Add a LazyPool test context and new SPDX fixtures to simulate many descendant SBOMs for pool exhaustion testing.

### CODE REVIEW

File: modules/analysis/src/service/test/external_depth.rs [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    config::AnalysisConfig,
+    model,
+    service::{AnalysisService, ComponentReference, QueryOptions},
+};
+use test_context::test_context;
+use trustify_common::model::Paginated;
+use trustify_test_context::TrustifyContext;
+
+/// Compute the maximum depth of the descendant tree.
+fn max_descendant_depth(nodes: &[model::Node]) -> u64 {
```
Comment (rh-jfuller) [inline modules/analysis/src/service/test/external_depth.rs:11]: nice ;)

**Review by rh-jfuller — APPROVED** [review]
LGTM ... there are some minor corner cases (that became apparent during review) related to dedup and cycle detection but they have always been there and I do not think impacts.

**Comment by rh-jfuller** [comment]
note: scale-tests use datasets that are prob tied to main (and will not work against release branch)

**Comment by ctron** [comment]
The idea is to merge this onto main. Including the changes from #2407 ... I'll leave this PR open, and create a new one, targeting main with the other changes as well.

**Comment by ctron** [comment]
I think we can close this one, as it's merged to `main`.
